### PR TITLE
support for downloads CDNs #4327

### PR DESCRIPTION
**Pull request by HBcao233**
I made this PR to fix this issue:
#4327

### CODE REVIEW

File: telethon/client/downloads.py [hunk 1]
```diff
@@ -27,39 +27,49 @@
...
     async def _init(
-            self, file, dc_id, offset, stride, chunk_size, request_size, file_size, msg_data
-    ):
+            self, file, dc_id, offset, stride, chunk_size, request_size, file_size, msg_data, cdn_redirect=None):
         self.request = functions.upload.GetFileRequest(
-            file, offset=offset, limit=request_size)
-
+            file, offset=offset, limit=request_size) 
+        self._client = self.client
+        self.cdn_redirect = cdn_redirect
```
Comment (Lonami) [inline telethon/client/downloads.py]: Is it intentional that this is public?

File: telethon/client/downloads.py [hunk 2]
```diff
@@ -73,10 +83,15 @@ async def _load_next_chunk(self):
 
     async def _request(self):
         try:
-            result = await self.client._call(self._sender, self.request)
+            result = await self._client._call(self._sender, self.request)
             self._timed_out = False
             if isinstance(result, types.upload.FileCdnRedirect):
-                raise NotImplementedError  # TODO Implement
+                # raise NotImplementedError  # TODO Implement
```
Comment (Lonami) [inline telethon/client/downloads.py]: Could delete this line now.

File: telethon/client/downloads.py [hunk 3]
```diff
@@ -73,10 +83,15 @@ async def _load_next_chunk(self):
...
+                raise _CdnRedirect(result)
+            if isinstance(result, types.upload.CdnFileReuploadNeeded):
+                result = await self._client._call(self._sender, functions.upload.reuploadCdnFile(file_token=self.cdn_redirect.file_token, request_token=result.request_token))
```
Comment (Lonami) [inline telethon/client/downloads.py]: Is this correct? Shouldn't it be `ReuploadCdnFileRequest`?

File: telethon/client/downloads.py [hunk 4]
```diff
@@ -561,6 +578,20 @@ async def _download_file(
 
             if in_memory:
                 return f.getvalue()
+        except _CdnRedirect as e:
+          self._log[__name__].info('FileCdnRedirect')
```
Comment (Lonami) [inline telethon/client/downloads.py]: Could print more info.

File: telethon/client/telegrambaseclient.py [hunk 5]
```diff
@@ -890,7 +892,8 @@ async def _clean_exported_senders(self: 'TelegramClient'):
     async def _get_cdn_client(self: 'TelegramClient', cdn_redirect):
         """Similar to ._borrow_exported_client, but for CDNs"""
         # TODO Implement
-        raise NotImplementedError
+        # raise NotImplementedError
```
Comment (Lonami) [inline telethon/client/telegrambaseclient.py]: Can delete these comments.

File: telethon/client/telegrambaseclient.py [hunk 6]
```diff
@@ -890,7 +892,8 @@ async def _clean_exported_senders(self: 'TelegramClient'):
...
+        from .telegramclient import TelegramClient
```
Comment (Lonami) [inline telethon/client/telegrambaseclient.py]: This seems quite hacky, but eh, as long as it works, I don't really care much in v1.

**Comment by Lonami** [comment]
Thanks.

**Comment by AmarnathCJD** [comment]
> I made this PR to fix this issue: #4327

Hi do you have link to a CDN redirected file.?

**Pull request merged**